lib/SILOptimizer/Transforms/SILSROA.cpp

```diff
@@ -260,12 +265,32 @@ void SROAMemoryUseAnalyzer::chopUpAlloca(std::vector<AllocStackInst *> &Worklist
 
   // Change any aggregate stores into extracts + field stores.
   for (auto *SI : Stores) {
-    SILBuilderWithScope B(SI);
-    for (unsigned EltNo : indices(NewAllocations))
-      B.createStore(SI->getLoc(),
-                    createAggProjection(B, SI->getLoc(), SI->getSrc(), EltNo),
-                    NewAllocations[EltNo],
-                    StoreOwnershipQualifier::Unqualified);
+    SILBuilderWithScope builder(SI);
+    // In ownership mode, create a destructure of the source and use that to
+    // initialize the aggregate.
+    MultipleValueInstruction *destructured = nullptr;
+    if (SI->getFunction()->hasOwnership()) {
```

Comment:
Rather than do it this way, there are helpers for destructures that work regardless of ossa or non-ossa. You want to use:

https://github.com/apple/swift/blob/7d2f9440866c5360cc4929f003850058ecfc98bc/include/swift/SIL/SILBuilder.h#L1470